modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/wal/FileWriteAheadLogManager.java

```diff
@@ -1261,24 +1284,38 @@ private FileWriteHandle rollOver(FileWriteHandle cur, @Nullable WALRecord rec) t
                 switchSegmentRecordOffset.set(idx, hnd.getSwitchSegmentRecordOffset());
             }
 
+            if (archiver == null)
+                segmentAware.addReservedWalArchiveSize(maxWalSegmentSize);
+
             FileWriteHandle next;
             try {
-                next = initNextWriteHandle(cur);
-            }
-            catch (IgniteCheckedException e) {
-                //Allow to avoid forever waiting in other threads.
-                cur.signalNextAvailable();
+                try {
+                    next = initNextWriteHandle(cur);
+                }
+                catch (IgniteCheckedException e) {
+                    //Allow to avoid forever waiting in other threads.
+                    cur.signalNextAvailable();
 
-                throw e;
-            }
+                    throw e;
+                }
 
-            if (rec != null) {
-                WALPointer ptr = next.addRecord(rec);
+                if (rec != null) {
+                    WALPointer ptr = next.addRecord(rec);
 
-                assert ptr != null;
+                    assert ptr != null;
+                }
+
+                segmentSize.put(next.getSegmentId(), maxWalSegmentSize);
+
+                if (archiver == null)
```

Comment:
Can we avoid archiver check somehow? Maybe we can hold this info inside of segmentAware?